// ng2-date-picker/esm5/day-time-calendar/day-time-calendar.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import * as momentNs from 'moment';
import { UtilsService } from '../common/services/utils/utils.service';
import { DayCalendarService } from '../day-calendar/day-calendar.service';
import { TimeSelectService } from '../time-select/time-select.service';
var moment = momentNs;
var DAY_FORMAT = 'YYYYMMDD';
var TIME_FORMAT = 'HH:mm:ss';
var COMBINED_FORMAT = DAY_FORMAT + TIME_FORMAT;
var DayTimeCalendarService = /** @class */ (function () {
    function DayTimeCalendarService(utilsService, dayCalendarService, timeSelectService) {
        this.utilsService = utilsService;
        this.dayCalendarService = dayCalendarService;
        this.timeSelectService = timeSelectService;
        this.DEFAULT_CONFIG = {
            locale: moment.locale()
        };
    }
    DayTimeCalendarService.prototype.getConfig = function (config) {
        var _config = tslib_1.__assign({}, this.DEFAULT_CONFIG, this.timeSelectService.getConfig(config), this.dayCalendarService.getConfig(config));
        moment.locale(config.locale);
        return _config;
    };
    DayTimeCalendarService.prototype.updateDay = function (current, day, config) {
        var time = current ? current : moment();
        var updated = moment(day.format(DAY_FORMAT) + time.format(TIME_FORMAT), COMBINED_FORMAT);
        if (config.min) {
            var min = config.min;
            updated = min.isAfter(updated) ? min : updated;
        }
        if (config.max) {
            var max = config.max;
            updated = max.isBefore(updated) ? max : updated;
        }
        return updated;
    };
    DayTimeCalendarService.prototype.updateTime = function (current, time) {
        var day = current ? current : moment();
        return moment(day.format(DAY_FORMAT) + time.format(TIME_FORMAT), COMBINED_FORMAT);
    };
    DayTimeCalendarService = tslib_1.__decorate([
        Injectable(),
        tslib_1.__metadata("design:paramtypes", [UtilsService,
            DayCalendarService,
            TimeSelectService])
    ], DayTimeCalendarService);
    return DayTimeCalendarService;
}());
export { DayTimeCalendarService };
